RHCLOUD-4809 Documentation of Pipeline/Deployment Process for Inventory (#634)
RedHatInsights/insights-host-inventory: README.md

Cron jobs push their metrics to a
[Prometheus Pushgateway](https://github.com/prometheus/pushgateway/) instance
running at _PROMETHEUS_PUSHGATEWAY_. Defaults to _localhost:9091_.

# Release process

This section describes the process of getting a code change from a pull request all the way to production.

## 1. Pull request

It all starts with a [pull request](https://github.com/RedHatInsights/insights-host-inventory/pulls).
When a new pull request is opened the following checks are [run automatically](https://github.com/RedHatInsights/insights-host-inventory/blob/master/Jenkinsfile):

> * [database migrations](#initialize-the-database)

Reviewer: I don't think linking this section is quite what we want. We should have a section in this README that discusses best practices for staging a DB migration. I imagine there is a "Flasky" way to do that?

Author: Not sure what you mean by "staging a db migration"

Reviewer: creating / testing one in local dev; i.e., getting it ready before the actual deploy to CI/QA env

Author: I think the closest we have is https://github.com/RedHatInsights/insights-host-inventory/tree/ec51004fc180cf74e87d3f3ae346cb94a983a124#generating-a-migration-script which describes how to create a new migration?

Reviewer: Precisely, yeah. I was suggesting that https://github.com/RedHatInsights/insights-host-inventory/tree/ec51004fc180cf74e87d3f3ae346cb94a983a124#generating-a-migration-script would be better place to link to here rather than the instructions on how to init the DB.

* [code style checks](#contributing)
* unit tests

Should any of these fail this is indicated directly on the pull request.

When all of these checks pass and a reviewer approves the changes the pull request can be merged.
Currently [@Glutexo](https://github.com/Glutexo) and [@jharting](https://github.com/jharting) are authorized to merge pull requests.

## 2. Latest image and smoke tests

When a pull request is merged to master a new container image is built and tagged as [insights-inventory:latest](https://console.insights-dev.openshift.com/console/project/buildfactory/browse/images/insights-inventory/latest?tab=body).

> The latest image is deployed to the [CI environment](https://console.insights-dev.openshift.com/console/project/platform-ci/overview) automatically. This is announced in slack as `Inventory release pipeline (step 1)`

Reviewer: do we already have these slack notifications setup? if so, what channel do they go to (it's worth mentioning the channel here too).

Author: Thanks for feedback. I added the slack channel reference

Afterwards, the smoke tests are started automatically. These consist of a subset of [iqe tests](https://gitlab.cee.redhat.com/insights-qe) that cover inventory, advisor, engine and the ingress service.

> Once the smoke test run finishes the results are reported in slack as `Inventory release pipeline (step 2)`

Reviewer: ditto RE: mentioning the slack channel

running at _PROMETHEUS_PUSHGATEWAY_. Defaults to _localhost:9091_.

# Release process

This section describes the process of getting a code change from a pull request all the way to production.

## 1. Pull request

It all starts with a [pull request](https://github.com/RedHatInsights/insights-host-inventory/pulls).
When a new pull request is opened the following checks are [run automatically](https://github.com/RedHatInsights/insights-host-inventory/blob/master/Jenkinsfile):

* [database migrations](#initialize-the-database)
* [code style checks](#contributing)
* unit tests

Should any of these fail this is indicated directly on the pull request.

When all of these checks pass and a reviewer approves the changes the pull request can be merged.
Currently [@Glutexo](https://github.com/Glutexo) and [@jharting](https://github.com/jharting) are authorized to merge pull requests.

## 2. Latest image and smoke tests

When a pull request is merged to master a new container image is built and tagged as [insights-inventory:latest](https://console.insights-dev.openshift.com/console/project/buildfactory/browse/images/insights-inventory/latest?tab=body).

> The latest image is deployed to the [CI environment](https://console.insights-dev.openshift.com/console/project/platform-ci/overview) automatically. This is announced in [#platform-inventory-standup slack channel](https://app.slack.com/client/T026NJJ6Z/CQFKM031T) as `Inventory release pipeline (step 1)`

Reviewer 2: ```suggestion
The latest image is deployed to the [CI environment](https://console.insights-dev.openshift.com/console/project/platform-ci/overview) automatically. This is announced in [#platform-inventory-standup slack channel](https://app.slack.com/client/T026NJJ6Z/CQFKM031T) as _Inventory release pipeline (step 1)_.
```

> Afterwards, the smoke tests are started automatically. These consist of a subset of [iqe tests](https://gitlab.cee.redhat.com/insights-qe) that cover inventory, advisor, engine and the ingress service.

Reviewer 2: ```suggestion
Afterwards, the smoke tests are started automatically. These consist of a subset of [IQE tests](https://gitlab.cee.redhat.com/insights-qe) that cover inventory, advisor, engine and the ingress service.
```

> Once the smoke test run finishes the results are reported in [#platform-inventory-standup slack channel](https://app.slack.com/client/T026NJJ6Z/CQFKM031T) as `Inventory release pipeline (step 2)`

Reviewer 2: ```suggestion
Once the smoke test run finishes the results are reported in [#platform-inventory-standup slack channel](https://app.slack.com/client/T026NJJ6Z/CQFKM031T) as _Inventory release pipeline (step 2)_.
```

> If the smoke tests pass the container image is re-tagged as [insights-inventory:stable](https://console.insights-dev.openshift.com/console/project/buildfactory/browse/images/insights-inventory/stable?tab=body)

Reviewer 2: ```suggestion
If the smoke tests pass, the container image is re-tagged as [insights-inventory:stable](https://console.insights-dev.openshift.com/console/project/buildfactory/browse/images/insights-inventory/stable?tab=body)
```

### Suppressing automatic propagation to stable

In certain situations it may be desired for an image to stay deployed in CI only for some time before proceeding with the QA deployment.
An example would be a possibly breaking change where we want to give integrators some time to test the integration in CI first.

> This can be achieved by setting the `promote-stable` key to `false` in the [inventory-pipeline config map](https://console.insights-dev.openshift.com/console/project/buildfactory/browse/config-maps/inventory-pipeline).

Reviewer 2: This looks like a bit cumbersome way to do it. But this pull request just describes the state, so ok.

Author: Let me know if you have ideas about a more natural way to control this setting

This change needs to be made **before the given commit is merged to master**.

## 3. Vortex

[Vortex](https://github.com/RedHatInsights/e2e-deploy/blob/master/docs/pipeline.md#overview) is a process that picks stable images of platform components and tests them together. The process is automated and runs periodically.
By default the [vortex job](https://jenkins-insights-qe-ci.cloud.paas.psi.redhat.com/blue/organizations/jenkins/vortex-test-suite/activity) runs around every two hours.

> Once the vortex tests are finished the stable inventory image is tagged as [insights-inventory:qa](https://console.insights-dev.openshift.com/console/project/buildfactory/browse/images/insights-inventory/qa?tab=body) and deployed to the [QA environment](https://console.insights-dev.openshift.com/console/project/platform-qa/overview). This is announced in [#platform-inventory-standup slack channel](https://app.slack.com/client/T026NJJ6Z/CQFKM031T) as `Inventory release pipeline (step 3)`.

Reviewer 2: ```suggestion
Once the vortex tests are finished the stable inventory image is tagged as [insights-inventory:qa](https://console.insights-dev.openshift.com/console/project/buildfactory/browse/images/insights-inventory/qa?tab=body) and deployed to the [QA environment](https://console.insights-dev.openshift.com/console/project/platform-qa/overview). This is announced in [#platform-inventory-standup slack channel](https://app.slack.com/client/T026NJJ6Z/CQFKM031T) as _Inventory release pipeline (step 3)_.
```

## 4. QE testing

> Once the stable image lands in the QA environment the QE testing can begin.

Reviewer 2: ```suggestion
Once the stable image lands in the QA environment, the QE testing can begin.
```

> [@platform-inventory-qe](https://app.slack.com/client/T026NJJ6Z/browse-user-groups/user_groups/S011SJB6S5R) handles this part. They triggers the QE testing jobs \[[1](https://jenkins-jenkins.5a9f.insights-dev.openshiftapps.com/view/QE/job/qe/job/iqe-inventory-plugin-kafka-qa/)\]\[[2](https://insights-stg-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/Inventory/job/iqe-host-inventory-qa-basic/)\]. Optionally, manual testing may be performed to validate certain features / bug fixes.

Reviewer 2: ```suggestion
People in [@platform-inventory-qe](https://app.slack.com/client/T026NJJ6Z/browse-user-groups/user_groups/S011SJB6S5R) Slack group handle this part. They trigger the QE testing jobs \[[1](https://jenkins-jenkins.5a9f.insights-dev.openshiftapps.com/view/QE/job/qe/job/iqe-inventory-plugin-kafka-qa/)\]\[[2](https://insights-stg-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/Inventory/job/iqe-host-inventory-qa-basic/)\]. Optionally, manual testing may be performed to validate certain features / bug fixes.
```

> Once all of this is finished [@platform-inventory-qe](https://app.slack.com/client/T026NJJ6Z/browse-user-groups/user_groups/S011SJB6S5R) reports the results back in the [#platform-inventory-standup channel](https://app.slack.com/client/T026NJJ6Z/CQFKM031T).

Reviewer 2: Automatically, or by coming and saying?

```suggestion
Once all of this is finished, [@platform-inventory-qe](https://app.slack.com/client/T026NJJ6Z/browse-user-groups/user_groups/S011SJB6S5R) reports the results back in the [#platform-inventory-standup channel](https://app.slack.com/client/T026NJJ6Z/CQFKM031T).
```

## 5. Promoting the image to production

> Once the image has passed testing it can be promoted to production.

Reviewer 2: ```suggestion
Once the image has passed QE testing, it can be promoted to production.
```

> This is done using a [Jenkins job](https://jenkins-insights-jenkins.1b13.insights.openshiftapps.com/job/platform-prod/job/platform-prod-insights-inventory-deployer/build).

Reviewer 2: I couldn’t take a look as I don’t have privileges for production.

Reviewer: @Glutexo I _think_ you just need to open a RHIOPS ticket asking for permissions OR if you just login with your openshift creds it might just work. Let me know if the latter doesn't work and I can help with the former.

> The sha of the HEAD commit should be used as `PROMO_CODE`.

Reviewer 2: ```suggestion
The SHA of the HEAD commit should be used as `PROMO_CODE`.
```

> A new deployment is announced in [#platform-inventory-standup slack channel](https://app.slack.com/client/T026NJJ6Z/CQFKM031T) as `Inventory release pipeline (step 4)`

Reviewer 2: ```suggestion
A new deployment is announced in [#platform-inventory-standup slack channel](https://app.slack.com/client/T026NJJ6Z/CQFKM031T) as _Inventory release pipeline (step 4)_.
```

I‘d love to have these steps described instead of just numbered. But again, this is not in the scope of this PR.
